cocotb/share/lib/verilator/verilator.cpp

```diff
@@ -56,44 +75,68 @@ int main(int argc, char** argv) {
 #endif
 #endif
 
+    bool again;
     while (!Verilated::gotFinish()) {
-        bool again = true;
+        // Call registered timed callbacks (e.g. clock timer)
+        // These are called at the beginning of the time step
+        // before the iterative regions (IEEE 1800-2012 4.4.1)
+        VerilatedVpi::callTimedCbs();
+
+        // Call Value Change callbacks triggered by Timer callbacks
+        // These can modify signal values
+        settle_value_callbacks();
 
         // We must evaluate whole design until we process all 'events'
+        again = true;
         while (again) {
             // Evaluate design
             top->eval();
 
-            // Call Value Change callbacks as eval()
-            // can modify signals values
-            VerilatedVpi::callValueCbs();
+            // Call Value Change callbacks triggered by eval()
+            // These can modify signal values
+            again = settle_value_callbacks();
 
-            // Call registered Read-Write callbacks
-            again = VerilatedVpi::callCbs(cbReadWriteSynch);
+            // Call registered ReadWrite callbacks
+            again = (again || VerilatedVpi::callCbs(cbReadWriteSynch));
```

Comment:
BTW clangtidy would probably complain that you can simplify "again = (again | ...":

```suggestion
            again |= VerilatedVpi::callCbs(cbReadWriteSynch);
```